add-friends-likes

ilya-noize opened the pull request.

ilya-noize committed:
Pull Request #2
add friends in users;
add likes in films;

ilya-noize committed:
Pull Request #2
add friends in users;
add likes in films;

ilya-noize committed:
Pull Request #2
add friends in users;
add likes in films;

ilya-noize committed:
Pull Request #2
add friends in users;
add likes in films;

ilya-noize committed:
Pull Request #2
add friends in users;
add likes in films;

ilya-noize committed:
Pull Request #2
add friends in users;
add likes in films;

ilya-noize committed:
Pull Request #2
add friends in users;
add likes in films;

ilya-noize committed:
Pull Request #2
add friends in users;
add likes in films;

ilya-noize committed:
Pull Request #2
add friends in users;
add likes in films;

ilya-noize committed:
Pull Request #2
add friends in users;
add likes in films;

friends in memory;

ilya-noize committed:
Pull Request #2
add friends in users;
add likes in films;

refactoring storages and services;
add something exceptions;

ilya-noize committed:
Pull Request #2
add friends in users;
add likes in films;

ilya-noize committed:
Pull Request #2
add friends in users;
add likes in films;

ilya-noize committed:
Pull Request #2

fix checkstyle;

ilya-noize committed:
Pull Request #2

fix checkstyle;

ilya-noize committed:
Pull Request #2

fix tests;

ilya-noize committed:
Pull Request #2

fix constructors;
renaming;

ilya-noize committed:
Pull Request #2

POSTMAN Film tests Ok.

ilya-noize committed:
Pull Request #2

Delete Controller.java
patch for test postman "User id=2 get 2 friends" (not worked)

ilya-noize committed:
Pull Request #2

fix checkstyle;

ilya-noize committed:
Pull Request #2

fix checkstyle;

ilya-noize committed:
Pull Request #2

friends in User.class

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/constraint/LoginWith.java:
```diff
@@ -14,9 +14,9 @@
     String value() default "qwertyuiopasdfghjklzxcvbnm1234567890-_.";
 
-    String message() default "login содержит некорректные символы";
+//    String message() default "login содержит некорректные символы";
```
Надо либо вернуть использование данной аннотации, либо стоит совсем от нее отказаться и перейти к иному способу валидации логина

ilya-noize replied:
Согласно тз такая проверка не нужна. Поэтому откажусь.

KirPyt replied:
Стоит тогда убрать данный код, чтобы он не мозолил глаз)

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/ErrorController.java:
```diff
@@ -0,0 +1,54 @@
+public class ErrorController {
+
+    @ExceptionHandler
+    @ResponseStatus(HttpStatus.OK)
```
Данный статус характеризует корректное выполнение действия, и для ошибки стоит использовать именно статусы ошибок - `BAD_REQUEST`, `NOT_FOUND`, и т.д

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/ErrorController.java:
```diff
@@ -0,0 +1,54 @@
+    @ExceptionHandler
+    @ResponseStatus(HttpStatus.OK)
+    public ErrorResponse handleEmptyFriendsSetException(EmptyFriendsSetException e) {
+        return new ErrorResponse(e.getMessage());
```
В каждый хендлер надо добавить логирование стек-трейса ошибки

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/ErrorController.java:
```diff
@@ -0,0 +1,54 @@
+
+    @ExceptionHandler
+    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
+    public ErrorResponse handleWrongFilmIdException(WrongFilmIdException e) {
```
В данном хендлере ловим `Throwable`

ilya-noize replied:
Для более глубокого анализа ошибки?

KirPyt replied:
Скорее для перехвата всех остальных исключений, хендлеры для которых явно не указаны, так как данный тип ошибки определяет самое большое множество исключений и ловит всё, что не обрабатывается явно

KirPyt replied:
Стоит заметить, что у нас должен быть именно хендлер, который ловит `Throwable`, а возвращает `Internal Server Error` как статус хендлера

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,58 @@
+    private final FilmService filmService;
+
+    @Autowired
+    FilmController(FilmService filmService) {
```
Если хотите, можно заменить конструктор на аннотацию Lombok:
`@RequiredArgsConstructor`

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,58 @@
+
+    @PostMapping
+    public Film create(@RequestBody @Valid Film film) {
+        return filmService.create(film);
```
В каждый метод контроллера стоит добавить логирование входящего запроса

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    @PutMapping("/{id}/like/{userId}")
+    public void addLike(@PathVariable String id, @PathVariable String userId) {
```
Здесь можно сразу ловить `long` вместо `String`, так как `spring` сделает всю остальную работу по преобразованию за Вас

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    @GetMapping({"/popular?count={count}", "/popular"})
+    public List<Film> getPopular(@RequestParam(defaultValue = "10", required = false) String count) {
```
При наличии `defaultValue` в `required = false` нет смысла)

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    @DeleteMapping("/{id}/like/{userId}")
+    public void deleteLike(@PathVariable String id, @PathVariable String userId) {
```
Здесь и далее аналогично, можно сразу ловить `long`

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,64 @@
+@RestController
+@Slf4j
+@RequestMapping("/users")
+public class UserController {
```
Такие же комментарии, что и для контроллера фильмов)

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/ErrorResponse.java:
```diff
@@ -0,0 +1,13 @@
+public class ErrorResponse {
+    private final String error;
+
+    public ErrorResponse(String error) {
```
Вы можете воспользоваться аннотациями от lombok - `@RequiredArgsConstructor` и `@Getter`, это позволит сократить код в данном классе

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -11,8 +13,7 @@
+@EqualsAndHashCode
 @ToString
 public class Film {
     private static final String RELEASE_DATE_LIMIT = "1895-12-28";
```
Можно сразу создать объект `LocalDateTime` и инициализировать его данной датой

ilya-noize replied:
Можно. Только IDEA не позволяет работать @CorrectReleaseDate с таким полем
/ Attribute value must be constant
/ Предлогает convert to ThreadLocal. Получается что-то страшное. Потому и использовал String.

ilya-noize replied:
RELEASE_DATE_LIMIT итак final. Может быть нужно что-то другое )

KirPyt replied:
В целом - нет, но это странно что нельзя сразу создать объект `LocalDate.of(1895, 12, 28)`

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -11,8 +13,7 @@
 @Data
-@AllArgsConstructor
-@NoArgsConstructor
+@EqualsAndHashCode
```
Данная аннотация уже есть в аннотации `@Data`

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -11,8 +13,7 @@
-@AllArgsConstructor
-@NoArgsConstructor
+@EqualsAndHashCode
 @ToString
```
Данная аннотация так же уже присутствует в `@Data`

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -9,12 +9,12 @@
+import java.util.Set;
 
 @Data
 @ToString
```
Данная аннотация так же уже присутствует в `@Data`

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmServiceImpl.java:
```diff
@@ -0,0 +1,186 @@
+@Slf4j
+@Service
+public class FilmServiceImpl implements FilmService {
+    private final UserService userService;
```
Рекомендую стараться использовать на сервисном слое только репозитории

ilya-noize replied:
Не знаком с @Repository толком. Это что-то вроде @Component, только с нюансами. Какими именно я пока не в курсе. Вроде как потом будет в курсе.

KirPyt replied:
Я имел в виду использовать здесь `FilmStorage` вместо `UserService`, на счёт аннотаций это, конечно, будет уже позже)

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmServiceImpl.java:
```diff
@@ -0,0 +1,186 @@
+    @Autowired
+    public FilmServiceImpl(FilmStorage filmStorage, UserStorage userStorage) {
+        this.filmStorage = filmStorage;
+        this.userService = new UserServiceImpl(userStorage);
```
Мы не создаем объекты через `new`, так как используем `@Autowired` - таким образом всю работу по созданию объектов за нас делает `spring`

ilya-noize replied:
В теоретической части я не нашёл упоминаний об этом.
Мог как-то пропустить материал к самостоятельному обучению?

KirPyt replied:
В целом если Вы запишите данный конструктор следующим образом:

```
@Autowired
public FilmServiceImpl(FilmStorage filmStorage, UserStorage userStorage) {
       this.filmStorage = filmStorage;
       this.userService = userStorage;
}
```

то всё должно работать

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmServiceImpl.java:
```diff
@@ -0,0 +1,186 @@
+        }
+        return getAll().stream()
+                .sorted((f1, f2) -> (
+                        filmStorage.getLikes(f2.getId()).size() - filmStorage.getLikes(f1.getId()).size()
```
Рекомендую вынести данный компаратор в константу

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmServiceImpl.java:
```diff
@@ -0,0 +1,186 @@
+        Integer filmId = film.getId();
+        Integer userId = user.getId();
+        Set<Integer> likes = filmStorage.getLikes(filmId);
+/*
```
Стоит убрать неиспользуемый код

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmServiceImpl.java:
```diff
@@ -0,0 +1,186 @@
+     */
+    private Film getFilmFromData(String supposedId) {
+        Integer filmId = integerFromString(supposedId);
+        if (filmId == Integer.MIN_VALUE || filmId <= 0) {
```
Рекомендую воспользоваться аннотацией `@PositiveOrZero` на уровне класса `Film`, тогда от проверки `filmId <= 0` можно будет отказаться

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,281 @@
+     */
+    private Set<Integer> findFriendsCommon(Set<Integer> friendsIdUser, Set<Integer> friendsIdUserOther) {
+        Set<Integer> friendsCommon = new HashSet<>();
+        for (Integer id : friendsIdUser) {
```
Стоит переписать все циклы с использованием `stream`, чтобы было удобнее воспользуйтесь методом `filter`, это отфильтровать вхождение элементов одного списка в другой

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,281 @@
+    private User getUserFromData(String supposedId) {
+        log.info("* Попытка получить данные пользователя");
+        Integer userId = integerFromString(supposedId);
+        if (userId == Integer.MIN_VALUE || userId <= 0) {
```
Аналогично, можно заменить проверку `userId <= 0` на аннотацию `@PositiveOrZero`

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
```diff
@@ -0,0 +1,56 @@
+
+@Component
+public class InMemoryFilmStorage implements FilmStorage {
+    Integer generateId;
```
Надо инкапсулировать данное поле

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,54 @@
+
+@Component
+@Slf4j
+public class InMemoryUserStorage implements UserStorage {
```
Рекомендую подумать над тем, чтобы сформировать абстрактное хранилище и вынести туда общую логику, так же в этом случае стоит воспользоваться дженериком - это позволит очень сильно сократить код в классах реализациях данного абстрактного хранилища

KirPyt commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
```diff
@@ -15,12 +22,13 @@
+    final int rightDuration = 1;
 
-    FilmController controller = new FilmController();
+    UserStorage userStorage = new InMemoryUserStorage();
```
Все поля надо инкапсулировать

KirPyt commented on src/test/java/ru/yandex/practicum/filmorate/controller/UserControllerTest.java:
```diff
@@ -14,7 +19,10 @@
     private static final LocalDate WRONG_BIRTHDAY = LocalDate.now();
-    UserController controller = new UserController();
+
+    UserStorage storage = new InMemoryUserStorage();
```
Все поля надо инкапсулировать

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,281 @@
+
+@Slf4j
+@Service
+public class UserServiceImpl implements UserService {
```
Рекомендую подумать над тем, чтобы сформировать абстрактный сервис и вынести туда общую логику для сервисов фильмов и пользователей, а затем отнаследоваться от абстрактного сервиса и таким образом получится сократить код

ilya-noize replied:
Это потянет за собой ещё и абстрактное Storage?

ilya-noize replied:
Из общих методов тут только integerFromString. И после перехода на Long он утратит свою необходимость.

KirPyt replied:
В целом сильно упростить тут не получится, только вынести примитивную логику - поиск объекта, создание, обновление и т.д - то что общее для сервисов фильмов и пользователей

KirPyt reviewed (commented):
Приветствую!

Указал некоторые замечания и советы по коду, предлагаю над ними поработать. Так же некоторые замечания или советы актуальны и для остальных частей кода, поэтому рекомендую обратить внимание и на другие аналогичные места в коде, чтобы всё поправить.

Если возникнут вопросы по ревью, то пишите мне в Пачку, обсудим)

ilya-noize committed:
Pull Request #2

after review

ilya-noize committed:
Pull Request #2

after review. remaster after night.

ilya-noize committed:
Merge remote-tracking branch 'origin/add-friends-likes' into add-friends-likes

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/ErrorController.java
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java
#	src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java
#	src/main/java/ru/yandex/practicum/filmorate/exception/EmptyFriendsSetException.java
#	src/main/java/ru/yandex/practicum/filmorate/model/Film.java
#	src/main/java/ru/yandex/practicum/filmorate/model/User.java
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmServiceImpl.java
#	src/main/java/ru/yandex/practicum/filmorate/service/UserService.java
#	src/main/java/ru/yandex/practicum/filmorate/service/UserServiceImpl.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/user/InMemoryUserStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/user/UserStorage.java
#	src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java
#	src/test/java/ru/yandex/practicum/filmorate/controller/UserControllerTest.java

ilya-noize committed:
Pull Request #2

after review. remaster after night-day.

ilya-noize committed:
Pull Request #2

after review. abstract class MainStorage<T>
try mockMvc

ilya-noize committed:
Pull Request #2
abstract class MainStorage<T> for Films, Users;
abstract class MainService<T> for Films, Users;

APPLICATION FAILED TO START
Description:
Parameter 1 of constructor in ru.yandex.practicum.filmorate.service.FilmLikesService required a bean of type 'ru.yandex.practicum.filmorate.service.MainService' that could not be found.

Action:
Consider defining a bean of type 'ru.yandex.practicum.filmorate.service.MainService' in your configuration.

ilya-noize committed:
Pull Request #2
abstract class MainStorage<T> for Films, Users;
abstract class MainService<T> for Films, Users;

APPLICATION FAILED TO START
Description:
Parameter 1 of constructor in ru.yandex.practicum.filmorate.service.FilmLikesService required a bean of type 'ru.yandex.practicum.filmorate.service.MainService' that could not be found.

Action:
Consider defining a bean of type 'ru.yandex.practicum.filmorate.service.MainService' in your configuration.

fix checkstyle;

ilya-noize committed:
Pull Request #2
abstract class MainStorage<T> for Films, Users;
abstract class MainService<T> for Films, Users;

APPLICATION FAILED TO START
Description:
Parameter 1 of constructor in ru.yandex.practicum.filmorate.service.FilmLikesService required a bean of type 'ru.yandex.practicum.filmorate.service.MainService' that could not be found.

Action:
Consider defining a bean of type 'ru.yandex.practicum.filmorate.service.MainService' in your configuration.

fix checkstyle;

ilya-noize committed:
Pull Request #2
abstract class MainStorage<T> for Films, Users;
abstract class MainService<T> for Films, Users;

APPLICATION FAILED TO START
Description:
Parameter 1 of constructor in ru.yandex.practicum.filmorate.service.FilmLikesService required a bean of type 'ru.yandex.practicum.filmorate.service.MainService' that could not be found.

Action:
Consider defining a bean of type 'ru.yandex.practicum.filmorate.service.MainService' in your configuration.

fix checkstyle;

ilya-noize committed:
Pull Request #2
abstract class MainStorage<T> for Films, Users;
abstract class MainService<T> for Films, Users;

APPLICATION FAILED TO START
Description:
Parameter 1 of constructor in ru.yandex.practicum.filmorate.service.FilmLikesService required a bean of type 'ru.yandex.practicum.filmorate.service.MainService' that could not be found.

Action:
Consider defining a bean of type 'ru.yandex.practicum.filmorate.service.MainService' in your configuration.

fix checkstyle;

ilya-noize committed:
Pull Request #2
abstract class MainStorage<T> for Films, Users;
abstract class MainService<T> for Films, Users;

APPLICATION FAILED TO START
Description:
Parameter 1 of constructor in ru.yandex.practicum.filmorate.service.FilmLikesService required a bean of type 'ru.yandex.practicum.filmorate.service.MainService' that could not be found.

Action:
Consider defining a bean of type 'ru.yandex.practicum.filmorate.service.MainService' in your configuration.

fix checkstyle;

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/ErrorController.java:
```diff
@@ -0,0 +1,47 @@
+
+    @ExceptionHandler
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    public ErrorResponse handleWrongUserIdException(WrongUserIdException e) {
```
Рекомендую не делать на каждый тип ошибки свой хендлер, лучше создать универсальное исключение - `ObjectNotFoundException` и ловить его

ilya-noize replied:
NotFoundException уже есть. Может быть это вообще лишнее. В коде я не нашёл проброс этого исключения.

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    @GetMapping({"/popular?count={count}", "/popular"})
+    public List<Film> getPopular(@RequestParam(defaultValue = "10", required = false) Long count) {
```
При наличии `defaultValue` писать `required = false` не нужно

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,62 @@
+@RequestMapping("/users")
+@RequiredArgsConstructor
+public class UserController {
+    @Autowired
```
Если делаете инжекцию через `@RequiredArgsConstructor`, то аннотацию `@Autowired` можно не писать

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmLikesService.java:
```diff
@@ -0,0 +1,91 @@
+        Film film = get(supposedId);
+        Set<Long> likes = film.getLikes();
+
+        User user = userService.storage.get(supposedUserId);
```
Объекты внутри других классов всегда получаем через get методы

ilya-noize replied:
Не совсем понял о чём это. Надо обдумать эту тему.

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,78 @@
+@Slf4j
+@Service
+public abstract class FilmService extends MainService<Film> {
+    protected final MainService<User> userService;
```
Стоит сделать данное поле `private` - так как иначе это порождает чрезмерную видимость объектов

ilya-noize replied:
Абстракцию MainService убрал совсем.

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/UserFriendsService.java:
```diff
@@ -0,0 +1,142 @@
+        Set<Long> friendsIdUser = user.getFriends();
+        Set<Long> friendsIdUserOther = userOther.getFriends();
+
+        return getFriendsSet(findFriendsCommon(friendsIdUser, friendsIdUserOther));
```
Рекомендую выносить промежуточные результаты в отдельные переменные - так проще читается

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/MainStorage.java:
```diff
@@ -0,0 +1,30 @@
+
+public abstract class MainStorage<T> {
+    private Long generateId = 1L;
+    protected final Map<Long, T> storage;
```
Аналогично, стоит сделать данное поле `private`

ilya-noize replied:
may be package-private? )) Трабл с доступом в сервисах. А поскольку модели разные, то доступ к разным полям разных моделей может быть невозможен.
Общие метода есть (getId/setId), но есть есть ещё setLikes/setFriends...
Возможно и Id будет общий на всех. Не проверял. Иду по пути наименьшего сопротивления.

ilya-noize replied:
Переименовал MainStorage в MasterStorage. Руки чесались это сделать ))

KirPyt commented on src/test/java/ru/yandex/practicum/filmorate/FilmorateApplicationTests.java:
```diff
@@ -1,12 +1,52 @@
+@RunWith(SpringRunner.class)
 @SpringBootTest
+@AutoConfigureMockMvc
+//@WebMvcTest (FilmController.class)
```
Неиспользуемый код стоит убрать

KirPyt commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
```diff
@@ -1,73 +1,72 @@
-    FilmController controller = new FilmController();
+    final LocalDate rightRelease = LocalDate.of(1895, 12, 28);
+    final int rightDuration = 1;
+    FilmController controller;
```
Надо инкапсулировать данное поле

KirPyt commented on src/test/java/ru/yandex/practicum/filmorate/controller/UserControllerTest.java:
```diff
@@ -1,26 +1,31 @@
     private static final LocalDate RIGHT_BIRTHDAY = LocalDate.of(2000, 1, 1);
     private static final LocalDate WRONG_BIRTHDAY = LocalDate.now();
-    UserController controller = new UserController();
+    UserController controller;
```
Надо инкапсулировать поле

KirPyt reviewed (commented):
Посмотрел работу - очень круто, что Вы решили реализовать абстрактный сервис и абстрактное хранилище - но, хочу заметить, что если будет сложно, то стоит отложить данную тему, так как она достаточно непростая.
Так же у Вас пока не прошли тесты на гитхабе (красный крестик у коммита), надо добиться прохождения тестов

ilya-noize committed:
Pull Request #2
abstract class MainStorage<T> for Films, Users;
abstract class MainService<T> deleted;
edits after the review:
 - 'required="10"' deleted;
 - encapsulating a variable;
 - rename MainStorage to MasterStorage
 - private final Map<Long, T> storage in UserStorage and FilmStorage

ilya-noize committed:
Pull Request #2
abstract class MasterService<T> for Films, Users remastered;

ilya-noize committed:
Pull Request #2
logging requests in create, update-methods
translate in english

ilya-noize committed:
Pull Request #2
In each handler added logging stack trace errors
translate in english

ilya-noize committed:
Pull Request #2
catch Throwable with HttpStatus.500

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,55 @@
+
+    @PutMapping("/{id}/like/{userId}")
+    public void addLike(@PathVariable Long id, @PathVariable Long userId) {
+        filmService.addLike(id, userId);
```
Стоит добавить недостающее логирование по аналогии с остальными методами

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/MasterService.java:
```diff
@@ -0,0 +1,31 @@
+
+import java.util.List;
+
+@Service
```
Данную аннотацию стоит указывать в непосредственной реализации сервиса, в двух местах её указывать не нужно

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/UserStorage.java:
```diff
@@ -0,0 +1,62 @@
+@Slf4j
+@RequiredArgsConstructor
+public class UserStorage extends MasterStorage<User> {
+    private final Map<Long, User> storage;
```
Думаю, можно было бы вынести данную переменную в абстрактное хранилище с типом `Map<Long, T>`

ilya-noize replied:
Можно, но private сменится на package-private или protected.  
Раз уж реализация методов в наследнике, то решил перенести Map в него.

KirPyt reviewed (commented):
В целом очень хорошо
Указал некоторые моменты, стоит их учесть на будущее, в остальном Ваш проект зачтён, хорошего дня!

ilya-noize committed:
Pull Request #2
Update README.md

ilya-noize committed:
Pull Request #2
edits after review. Merge branch with main
